CMakeLists.txt

```diff
@@ -84,69 +88,92 @@ file(GLOB_RECURSE CUDA_SOURCES
+list(APPEND LINK_LIBRARIES rabit::rabit)
 
 # dmlc-core
-add_subdirectory(dmlc-core)
-set(LINK_LIBRARIES dmlccore rabit)
+hunter_add_package(dmlc-core)
+find_package(dmlc CONFIG REQUIRED)
+list(APPEND LINK_LIBRARIES dmlc::dmlc)
 
+set(xgboost_exports "")
 
 if(USE_CUDA)
+  
   find_package(CUDA 8.0 REQUIRED)
   cmake_minimum_required(VERSION 3.5)
 
-  add_definitions(-DXGBOOST_USE_CUDA)
-  
-  include_directories(cub)
-
-  if(USE_NCCL)
-    include_directories(nccl/src)
-    add_definitions(-DXGBOOST_USE_NCCL)
+  list(APPEND xgboost_defs XGBOOST_USE_CUDA)  
+
+  # Compatibiltiy with CMake's CUDA_SELECT_NVCC_ARCH_FLAGS
+  # * All     : all CUDA architectures
+  # * Auto    : for the detected CUDA architecture only
+  # * Common  : all common CUDA architectures
+  set(xgboost_CUDA_VALID_TYPES Auto Common All Default Min)
+  set(xgboost_CUDA_TYPE Auto CACHE STRING "Cuda type (Auto|Common|All")
+  set_property(CACHE xgboost_CUDA_TYPE PROPERTY STRINGS ${xgboost_CUDA_VALID_TYPES})
+  if(NOT xgboost_CUDA_TYPE IN_LIST xgboost_CUDA_VALID_TYPES)
+    message(FATAL_ERROR "xgboost_CUDA_TYPE must be one of ${xgboost_CUDA_VALID_TYPES}")
   endif()
+  cuda_select_nvcc_arch_flags(GENCODE_FLAGS ${xgboost_CUDA_TYPE})
 
-  if((CUDA_VERSION_MAJOR EQUAL 9) OR (CUDA_VERSION_MAJOR GREATER 9))
-    message("CUDA 9.0 detected, adding Volta compute capability (7.0).")
-    set(GPU_COMPUTE_VER "${GPU_COMPUTE_VER};70")
-  endif()
+  # cub
+  hunter_add_package(cub)
+  find_package(cub CONFIG REQUIRED)
+  list(APPEND LINK_LIBRARIES cub::cub)
   
-  set(GENCODE_FLAGS "")
-  format_gencode_flags("${GPU_COMPUTE_VER}" GENCODE_FLAGS)
+  # Hard coding architectures can result in errors like:
+  # nvcc fatal   : Unsupported gpu architecture 'compute_70'
+  # if((CUDA_VERSION_MAJOR EQUAL 9) OR (CUDA_VERSION_MAJOR GREATER 9))
+  #   message("CUDA 9.0 detected, adding Volta compute capability (7.0).")
+  #   set(GPU_COMPUTE_VER "${GPU_COMPUTE_VER};70")
+  # endif()
+  #set(GENCODE_FLAGS "")
+  #format_gencode_flags("${GPU_COMPUTE_VER}" GENCODE_FLAGS)
+
   set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS};--expt-extended-lambda;--expt-relaxed-constexpr;${GENCODE_FLAGS};-lineinfo;")
   if(NOT MSVC)
-    set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS};-Xcompiler -fPIC; -std=c++11")
+    set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS};-Xcompiler -fPIC")
+    # Note adding -std=c++11 to CUDA_NVCC_FLAGS can cause duplicate flag errors
+    # See: https://gitlab.kitware.com/cmake/cmake/merge_requests/1628
+    # nvcc fatal   : redefinition of argument 'std'    
+    if(NOT "${CMAKE_CXX_FLAGS}" MATCHES "std=c\\+\\+") 
+      list(APPEND CUDA_NVCC_FLAGS "-std=c++11") # if language isn't specified already
+    endif()
   endif()
 
-  if(USE_NCCL)
-    add_subdirectory(nccl)
-  endif()
-
-  cuda_add_library(gpuxgboost ${CUDA_SOURCES} STATIC)
+  # Note: Some CMake generators won't support libraries without at least 1 source file:
+  set(xgboost_gpu_tmp ${CMAKE_CURRENT_BINARY_DIR}/tmp/xgboost_gpu.cc)
```

Comment:
@ruslo: I'm starting to use this as a common CMake pattern for `add_library` and `cuda_add_library` calls.  I believe it handles dependencies correctly (`copy_if_different`) and per documentation if the `OUTPUT` file is explicitly included in a target's sources it should all work out fine.  Let me know if you see any improvements, since it happens 2x here.  The Xcode generator (at least) requires at least one c++ file for both `cuds_add_library` calls (to complement a list of `*.cu` files) and for `OBJECT` libraries.

Reply:
Is is for Xcode limitation of at least one real file in sources? Why not just use some dummy file?

Reply:
It seemed to be less intrusive address this directly in CMakeLists.txt with something like `file(WRITE "" null.cc)`, but when the extra `add_custom_command` logic is added to prevent rebuilds, it probably is cleaner just to add empty files as needed.